### PR TITLE
[Bugfix][ROCm][Spec Decode] Fix defects on the Kimi-K3 DSpark decode path

### PR DESCRIPTION
**Pull request by JohnQinAMD**
## Purpose

Two defects on the Kimi-K3 DSpark decode path:

1. The draft model turns cudagraphs off for the whole target model.
2. The multi-token verify is split into one decode per token, which costs two waits on the GPU every forward.

### (1) A draft's attention decides the target's cudagraph mode

Before it captures anything, vLLM asks every attention group whether it can be graphed and takes the lowest answer. Groups that belong to a spec-decode draft are asked too:

```python
for group in groups:
    cg_support = builder.get_cudagraph_support(vllm_config, group.kv_cache_spec)
    if cg_support.value < min_cg_support.value:
        min_cg_support = cg_support          # draft-owned groups included
```

That lowest answer then decides the mode for the *target*:

```python
if (cudagraph_mode.decode_mode() == CUDAGraphMode.FULL
        and uniform_decode_query_len > 1
        and min_cg_support.value < AttentionCGSupport.UNIFORM_BATCH.value):
    cudagraph_mode = CUDAGraphMode.NONE      # PIECEWISE only if splitting_ops
```

Kimi-K3 meets all three conditions. The target runs AiterMLA, which reports `UNIFORM_BATCH`, and the recipe puts the draft on TritonMLA, which reports `UNIFORM_SINGLE_TOKEN_DECODE`. The draft's answer is the lower one, so **5 draft layers turn cudagraphs off for all 93 target layers**. The draft never needed that help: `init_cudagraph_manager` already looks at its own support level and runs the draft eagerly when it cannot be graphed.

**Fix: leave a draft that already checks itself out of the target's decision, and keep a second minimum for the check that still needs every group.**

Some drafts do not check themselves. Eagle, mtp and gemma4 (`AutoRegressiveSpeculator`), and `MultiModuleMTPSpeculator`, simply use whatever mode the target ended up with. For those the downgrade is doing real work: take it away and they would capture in a mode their own backend cannot run. So the test is a flag on the speculator, not its class:

```python
if isinstance(speculator, DraftModelSpeculator) and speculator.sizes_own_cudagraph_mode:
    return speculator.draft_attn_layer_names
```

`init_attn_backend` now takes two minimums instead of one. The old `min_cg_support` still covers every group, and adaptive verification reads it, so that check is unchanged. The new `graph_min_cg_support` leaves out groups made up entirely of draft layers, and only the cudagraph decision reads it. A group that mixes target and draft layers stays in both. Today only `DFlashSpeculator` and `DSparkSpeculator` set the flag.

### (2) The verify block is split into one decode per token

The draft proposes several tokens at once and the target checks them in one block — 8 of them with the `num_speculative_tokens=7` this recipe uses. Each token in the block may only look at the text before it, not after it. In bf16 there is no asm kernel for fewer than 16 heads with more than one query token, so a 12-head rank sends this block to Gluon instead. (With an fp8 cache the q-row fold — folding query rows into the head dimension — does give it an asm kernel, so only bf16 comes here.)

The old code faked the "look backward only" rule by issuing one single-token decode per verify token, each with its own slice of the KV list. Building those slices meant rewriting the paged-KV table on every forward, and twice asking the GPU for a value and waiting for it:

```python
new_indptr = torch.cat([old_indptr.new_zeros(1), row_len.cumsum(0)]).to(torch.int32)
total = int(new_indptr[-1].item())                  # host sync
...
new_indices = decode.paged_kv_indices[src]          # gather over the whole batch
mla_gluon(..., page_table=new_indices, seq_info=new_indptr,
          min_kv_seq_len=int(row_len.min()))        # host sync
```

**Fix: give Gluon the whole block at once, which it already accepts, and delete the splitting.**

Gluon takes a `[batch, qlen, nhead, dim]` query and applies the backward-only rule itself, so reshaping the query is all that is needed. The page table goes through untouched:

```python
q_nope = q_nope.view(batch_size, qlen, num_q_heads, self.kv_lora_rank)
q_pe = q_pe.view(batch_size, qlen, num_q_heads, self.qk_rope_head_dim)
mla_gluon(..., page_table=decode.paged_kv_indices, seq_info=decode.paged_kv_indptr,
          min_kv_seq_len=decode.min_kv_seq_len)
```

Net −71 lines in `rocm_aiter_mla.py`.

## Test Plan

```bash
pytest tests/v1/worker/test_attn_utils.py                       # 8, CPU only

pytest tests/kernels/attention/test_rocm_aiter_mla_causal_verify_mask.py   # 2, ROCm + AITER
```

The first needs no GPU: it drives `init_attn_backend` with stubs. The second builds
the real AITER backend and metadata on device and replaces the Gluon kernel with a
spy, so it skips off ROCm.

End to end on 8x MI355X, image `vllm/vllm-openai-rocm:nightly-aa9903490c616dc6871e5acc62cec7bb1e5e9434`, with `/model` = `moonshotai/Kimi-K3` and `/draft` its DSpark draft. Serve flags follow #51253:

```bash
vllm serve /model --served-model-name k3 \
  --tensor-parallel-size 8 --trust-remote-code \
  --kv-cache-dtype fp8 \
  --max-model-len 16384 --max-num-seqs 128 --max-num-batched-tokens 16384 \
  --gpu-memory-utilization 0.93 --block-size 128 --no-enable-prefix-caching \
  --load-format auto --mm-encoder-tp-mode data \
  --safetensors-load-strategy prefetch \
  --compilation-config '{"custom_ops":["+fused_rms_norm_gated"]}' \
  --speculative-config '{"method":"dspark","model":"/draft",
                         "num_speculative_tokens":7,
                         "attention_backend":"TRITON_MLA"}' \
  --port 8000
```

Run it twice: as written for the fp8 numbers, then with `--kv-cache-dtype fp8` dropped for the bf16 numbers. That one flag is what sends the verify to the padded asm decode or to Gluon.

```bash
for C in 1 8; do
  vllm bench serve --backend openai-chat --endpoint /v1/chat/completions \
    --model k3 --tokenizer /model --trust-remote-code --port 8000 \
    --dataset-name random --random-input-len 1024 --random-output-len 128 \
    --ignore-eos --max-concurrency $C --num-prompts $((C == 1 ? 16 : 48))
done

lm_eval --model local-chat-completions \
  --model_args model=k3,base_url=http://127.0.0.1:8000/v1/chat/completions,num_concurrent=32,tokenized_requests=False \
  --tasks gsm8k --num_fewshot 5 --batch_size 32 \
  --apply_chat_template --gen_kwargs temperature=0
```

`--apply_chat_template` and `tokenized_requests=False` are both needed against a chat endpoint. `--gen_kwargs temperature=0` matters because this model's `generation_config.json` sets no temperature, so the default is 1.0 and the run would otherwise be sampled.

## Test Result

Unit tests: **10 passed** (8 without a GPU, plus the 2 ROCm ones). For the end-to-end runs the control is this branch's base and the patched side is its head, in the same container, against the same aiter build.

Both dtypes are here because the KV cache dtype picks the decode kernel, and only one of them reaches the kernel (2) changes.

**fp8 sends the verify to the aiter asm decode**, and that is the production setup — the Kimi-K3 recipe runs `--kv-cache-dtype fp8`. Both of Gluon's checks turn a quantized cache away, so the verify falls through to the padded asm path: `get_mla_padded_q` pads 12 heads to 16, then `mla_decode_fwd`. That run covers (1), and the asm decode is the kernel that restored capture now has to hold.

**bf16 sends the verify to Gluon**, and it is the only setup that does, so it is the one that covers (2).

| KV cache | decode kernel | conc. | base | patched | speedup |
|---|---|---|---|---|---|
| fp8 | aiter asm (padded) | 1 | 14.82 tok/s, TPOT 66.09 ms | **67.41 tok/s, TPOT 13.44 ms** | **4.5x** |
| fp8 | aiter asm (padded) | 8 | 108.49 tok/s, TPOT 66.72 ms | **291.06 tok/s, TPOT 23.96 ms** | **2.7x** |
| bf16 | Gluon | 1 | 15.01 tok/s, TPOT 65.00 ms | **72.22 tok/s, TPOT 12.24 ms** | **4.8x** |
| bf16 | Gluon | 8 | 102.57 tok/s, TPOT 68.88 ms | **309.07 tok/s, TPOT 22.37 ms** | **3.0x** |

Both control runs report `cudagraph_mode=NONE`, 8 downgrade warnings and 0 captures. Both patched runs report 0 warnings and 8 captures, with the draft eager on all 8 ranks.

Accuracy, on the full GSM8K test set (1319 questions), 5-shot, greedy, fp8 KV:

| | flexible-extract | strict-match |
|---|---|---|
| base | 0.9613 ± 0.0053 | 0.9598 ± 0.0054 |
| patched | 0.9651 ± 0.0051 | 0.9644 ± 0.0051 |

Operator benchmark for (2), against Gluon directly: 12 heads, kv_lora 512, rope 64, bf16 KV. The splitting work is inside the timed region, because the old code did it on every forward. Over batch 8-32, qlen 2-8 and ctx 8k-32k, sending all the tokens at once is **1.52-4.07x faster** — 383.8 us down to 148.0 us at batch 8, qlen 8, ctx 8192. Outputs agree to 1.5e-05.

---

Written with AI assistance (Claude Code); the commits carry `Co-authored-by: Claude`. I reviewed every changed line, ran the unit tests and the end-to-end runs above myself, and am responsible for the result.

### CODE REVIEW

**Comment by tjtanaa** [comment]
@njhill @yewentao256 could you guys take alook at this PR?

**Review by yewentao256 — COMMENTED** [review]
I think @benchislett would be better person to review this PR

**Comment by JohnQinAMD** [comment]
Rebase on current main. 

One dependency worth flagging:  **This test needs [ROCm/aiter#4494](https://github.com/ROCm/aiter/pull/4494) landed before it is safe to merge.** Restoring graph capture re-exposes the asm split-K semaphore bug that PR fixes: the server serves normally, then wedges under concurrent load with every rank stuck on the same collective. #4494 was merged 2026-08-12 and reverted the same day by [#4709](https://github.com/ROCm/aiter/pull/4709) , so aiter main branch does not carry it today. I reproduced both the wedge and the fix [ROCm/aiter#4494](https://github.com/ROCm/aiter/pull/4494) is needed on this build.
